### PR TITLE
DataViews: Implement `isItemClickable` and `onClickItem` props

### LINKED ISSUE #66338: Data Views: Implement `selectItem` callback
During the review of the [Featured Image Control](https://github.com/WordPress/gutenberg/pull/64496#issuecomment-2429367097), the need for a `selectItem` callback was identified. This callback will enable layouts to trigger the item selection handler following user interactions.

### LINKED ISSUE #66588: Pages data view: Clicking the featured image in grid layout no longer opens the editor
On the Patterns and Templates data views, clicking the thumbnail in grid layout opens the editor. This used to be the case in the Pages data view too, but is no longer the case.

cc @oandregal @ntsekouras @youknowriad

### PR DESCRIPTION
**Pull request by gigitux**
## What?

This PR fixes https://github.com/WordPress/gutenberg/issues/66338.
This PR fixes https://github.com/WordPress/gutenberg/issues/66588.

This PR adds two props to the Data Views library:

### onClick

This callback is triggered when the user clicks on a media field or a primary field.

### isClickable

This function determines if a media field or a primary field are clickable. It takes an item as an argument and returns a boolean value indicating whether the item can be clicked.

## Why?

These two callbacks allow developers to implement click handlers for the media field and primary field at the consumer level, rather than within a specific field.

## Testing Instructions

Ensure that `Quick Edit in DataViews` and `Quick Edit in DataViews` are enabled.

1. Visit the Site Editor
2. Click on pages.
3. Toggle to the table view.
4. Ensure that when you click on the title you are redirected to the page editor
5. Go back.
6. Toggle to the grid view.
7. Ensure that when you click on the image, you are redirected to the page editor.

### CODE REVIEW

File: packages/dataviews/README.md [hunk 1]
```diff
@@ -320,6 +320,14 @@ The `defaultLayouts` property should be an object that includes properties named
 
 Callback that signals the user selected one of more items, and takes them as parameter. So far, only the `list` view implements it.
 
+### `isClickable`: `function`
```
Comment (gigitux) [inline packages/dataviews/README.md]: Not sure about the naming.

File: packages/dataviews/README.md [hunk 2]
```diff
@@ -320,6 +320,14 @@ The `defaultLayouts` property should be an object that includes properties named
...
+
+A function that determines if a media field or a primary field are clickable. It receives an item as an argument and returns a boolean value indicating whether the item can be clicked.
+
+### `onClick`: `function`
```
Comment (gigitux) [inline packages/dataviews/README.md]: Same here.

**Comment by youknowriad** [comment]
I think this is the right approach. Not sure I like the "click" terminology but I'm not the best with copy. Also curious what you think @oandregal.

**Comment by gigitux** [comment]
> I think this is the right approach

Thanks for the feedback! 💪 

> Not sure I like the "click" terminology

I attempted to find naming to make explicit that this callback is invoked exclusively for media and primary fields but struggled to find a suitably descriptive name. Feedback is welcome! 🙇

**Comment by oandregal** [comment]
I agree we need to make fields view-agnostic. The problem we need to solve is "Dynamic field props depending on layout": behaviours (onClick, onDrop, etc.) but also formats (sizes, link, etc.). I'm unconvinced this PR is the implementation we need to solve that problem.

- How this approach scales to 3rd party layouts? Adding props to the DataViews component for things our core layouts need would solve it for us. But, how can 3rd party layouts add the props they need? If we implemented calendar or kanban we'd have other feedback — I presume we're being misled by our core layouts being too similar. I think we need to open the layout API for 3rd parties so we can gather feedback from other kind of layouts.

- Additionally, this approach doesn't solve the [other aspect I mentioned](https://github.com/WordPress/gutenberg/pull/64496/files#r1813135126) about coupling fields&layout: providing different sizes.

I wonder if we could explore something tied to the layout definition. Perhaps the layout has the ability to add props to fields (mediaField, perhaps also address fields by type, etc.)? For example:

```js
{
  grid: {
    layout: {
      mediaField: {
        id: 'featured_media',
        onClick: 'action_id'
        size: {
          width: 32px,
          height: 32px
        }
      }
    }
  }
}
```

**Comment by youknowriad** [comment]
> How this approach scales to 3rd party layouts? Adding props to the DataViews component for things our core layouts need would solve it for us. But, how can 3rd party layouts add the props they need? If we implemented calendar or kanban we'd have other feedback — I presume we're being misled by our core layouts being too similar. I think we need to open the layout API for 3rd parties so we can gather feedback from other kind of layouts.

For me the solution in this PR is layout agnostic and can scale to any 3rd party layout. This is why I think the naming of the props is confusing. The say way `getItemId` is about retrieve the "id" of an item, this prop for me is about defining what it means to "navigate to a single item" or "focus on a single item". Third-party layouts can decide to use or not this "special action" (Should it be an action 🤔) when they want. 

Grid, list and table can decide to trigger it when you click or double click primary fields or the whole row...

> Additionally, this approach doesn't solve the [other aspect I mentioned](https://github.com/WordPress/gutenberg/pull/64496/files#r1813135126) about coupling fields&layout: providing different sizes.

Yes, this is not meant to address this, I think these two things are separate issues, for this one I suggested that `field.render` function should receive a "size" or "showDetails" props or things like that in the same way, layouts can decide to use or not. 

I do think there's probably a need to allow these to be customized within the "view" config in some situations but I believe in most situations layouts can enforce the "size" it wants (including third-parties);

**Comment by gigitux** [comment]
Upon further reflection, and after reading @oandregal’s comment, I find myself agreeing with @oandregal’s perspective. Primarily, I believe these kinds of actions are more related to the specific layout than to the dataviews.

We already have some confusion between some callback that we can pass to the dataviews:

https://github.com/WordPress/gutenberg/blob/5bf972645526b6d6ac1b206c8352a8e12d75c18d/packages/dataviews/README.md#L319-L329

- `onChangeSelection`: this callback works only for the `list` layout.
- `isClickable` and `onClick`: these callbacks work only for the `grid` and `list` layout.


I'm thinking if we should move these callbacks to the layout level, where each layout can define custom callback easily (as @oandregal suggested).

**Comment by youknowriad** [comment]
How is `onChangeSelection` specific to list? We can also select items in table and grid?

> - isClickable and onClick: these callbacks work only for the grid and list layout.

How is this a problem? The user of DataViews is defining what it means to "focus an item". Whether the layout (first or third-party) use this information or not is the responsibility of the layout component itself. 

I can imagine a "calendar" layout where double clicking a date focuses the post published on that date for instance...

**Comment by gigitux** [comment]
For example, currently this interaction of the `list` view is handled by the `onChangeSelection` callback:

https://github.com/user-attachments/assets/fec87e1b-959a-4290-9958-eb84fb6604f4


> Grid, list and table can decide to trigger it when you click or double click primary fields or the whole row...

As a consumer, I would expect that this interaction is handled by `onClick` (or `selectItem` or whatever name). IMHO, I'm afraid that consumers could be confused. 

For third-party developers, understanding which interaction triggers the callback can be complex. Moving these “settings” to the layout level can simplify this process IMHO.

**Comment by youknowriad** [comment]
> For example, currently this interaction of the list view is handled by the onChangeSelection callback:

What's important to note here, is that it's not the "list view" that is handling that click like that. It's the implementation of the list view in that particular page and that is totally fine. What the list view does is to change the "selection" and notifies the user, just like the other layouts.

> For third-party developers, understanding which interaction triggers the callback can be complex. Moving these “settings” to the layout level can simplify this process IMHO.

I think it's ok to be opinionated, I actually think it's a good thing to be opinionated. The layout is built in a way to be used like that, if you don't like it, build your own layout. I don't think we should optimize too much for flexibility here. I don't see us adding things like "onClick" or "onDoubleClick" or things like that to layouts. this is not semantic at all and can break with any redesign we make.

**Comment by gigitux** [comment]
> How is onChangeSelection specific to list? We can also select items in table and grid?

Thanks to your question, I realized that the documentation wasn't updated. I removed the `So far, only the `list` view implements it.` sentence: a8e3d7569e9252ed9eb5517247c00b39386baf2a

> I think it's ok to be opinionated, I actually think it's a good thing to be opinionated. The layout is built in a way to be used like that, if you don't like it, build your own layout

Gotcha!

File: packages/edit-site/src/components/post-list/index.js [hunk 3]
```diff
@@ -402,6 +400,14 @@ export default function PostList( { postType } ) {
 				onChangeView={ setView }
 				selection={ selection }
 				onChangeSelection={ onChangeSelection }
+				isClickable={ ( item ) => item.status !== 'trash' }
+				handleClick={ ( { id } ) => {
```
Comment (louwie17) [inline packages/edit-site/src/components/post-list/index.js]: I believe this should be `onClick` right?

**Review by louwie17 — CHANGES_REQUESTED** [review]
Thanks for this @gigitux, the testing instructions unfortunately didn't work. I think cause your using `handleClick` in the post list instead of `onClick`.
For the rest the code looks pretty straightforward. Just a couple thoughts on the overall implementation though:

If we go ahead with this, I think we should maybe change how the selection works? mostly cause we show the checkbox on hover, which to me gives the impression that click only does select.
Maybe there is some better way to highlight that clicking the title or image actually redirects to the site editor?
It would be great if we could use the `link` for this so that the browser highlights its actually a link.
These may be more design related questions cc: @jameskoster

File: packages/edit-site/src/components/post-list/index.js (hunk 3 above)
Comment (gigitux) [reply packages/edit-site/src/components/post-list/index.js]: Thanks for catching this! Fixed with c9ab19d095f3211c082a075472702df575938fe1

**Comment by youknowriad** [comment]
> If we go ahead with this, I think we should maybe change how the selection works? mostly cause we show the checkbox on hover, which to me gives the impression that click only does select.

I don't think the current PR should be about changing how things work. Selection should continue to work like trunk and clicking should only be applied to "primary fields"... not the whole row.

**Comment by louwie17** [comment]
> > If we go ahead with this, I think we should maybe change how the selection works? mostly cause we show the checkbox on hover, which to me gives the impression that click only does select.
> 
> I don't think the current PR should be about changing how things work. Selection should continue to work like trunk and clicking should only be applied to "primary fields"... not the whole row.

Yes I would agree with this, as it stands this PR does change this behaviour and the entire row ( except for the selection column ) redirects to the page site editor ( while in trunk it selects the row ).
Also the title isn't a link anymore so its hard to tell where we get redirected to, unlike before in trunk where the title is an actual link:
Screenshot 2024-11-04 at 11 37 54 AM

File: packages/dataviews/src/dataviews-layouts/table/index.tsx [hunk 4]
```diff
@@ -102,13 +110,32 @@ function TableColumnField< Item >( {
...
+				}
+			} }
+			onKeyDown={ ( event ) => {
+				if (
+					( event.key === 'Enter' || event.key === '' ) &&
+					isClickable( item )
+				) {
+					onClick( item );
+				}
+			} }
```
Comment (louwie17) [inline packages/dataviews/src/dataviews-layouts/table/index.tsx]: This breaks how the `onChangeSelection` works currently and causes `onClick` to be triggered for every column except for the selection column.

File: packages/edit-site/src/components/post-fields/index.js [hunk 5]
```diff
@@ -164,30 +159,10 @@ function usePostFields( viewType ) {
...
-						<Link
-							params={ {
-								postId: item.id,
-								postType: item.type,
-								canvas: 'edit',
-							} }
-						>
-							{ decodeEntities( renderedTitle ) ||
-								__( '(no title)' ) }
-						</Link>
```
Comment (louwie17) [inline packages/edit-site/src/components/post-fields/index.js:184]: Is there a way to still keep the actual link?

File: packages/dataviews/src/dataviews-layouts/table/index.tsx (hunk 4 above)
Comment (louwie17) [reply packages/dataviews/src/dataviews-layouts/table/index.tsx]: Could this be only added for the primary field? ( `primaryField?.id === field.id` ? )

**Review by louwie17 — CHANGES_REQUESTED** [review]
Thanks for updating the `onClick` callback, it did test correctly now according to your test instructions.
Although the behaviour in the table did change, as every column redirects to the site editor now, instead of just the primary column ( I left an inline comment regarding this ).

Also I think it would still be great if the primary fields could be displayed as a link if it actually redirects to a different page ( see image example in a previous comment above ).

**Comment by gigitux** [comment]
To enhance clarity, `isClickable` has been renamed to `isItemClickable` and `onClick` to `onItemClick` (f8d86b1f7d113f1f7f294effd6cb5eb2ebdd387b)

File: packages/dataviews/src/dataviews-layouts/table/index.tsx (hunk 4 above)
Comment (gigitux) [reply packages/dataviews/src/dataviews-layouts/table/index.tsx]: Thanks for catching that! I fixed the issue:

https://github.com/user-attachments/assets/d34dcfc3-c32c-4003-9860-59c30f94ee15


I wrapped the primary field in a dedicated div to ensure `onItemClick` is invoked only when the user clicks directly on the field.

https://github.com/WordPress/gutenberg/blob/5c7b0c45c8f68f863521f3312c161d8b3a8d9d83/packages/dataviews/src/dataviews-layouts/table/index.tsx#L124-L144

File: packages/edit-site/src/components/post-fields/index.js (hunk 5 above)
Comment (gigitux) [reply packages/edit-site/src/components/post-fields/index.js:184]: https://github.com/WordPress/gutenberg/blob/5c7b0c45c8f68f863521f3312c161d8b3a8d9d83/packages/dataviews/src/dataviews-layouts/table/index.tsx#L124-L146

We could wrap the field render in a `Link` component, but I'm not sure. The primary field might contain another `a` element, which could trigger a `ValidateDomNesting` warning.

File: packages/dataviews/README.md [hunk 6]
```diff
@@ -318,7 +318,15 @@ The `defaultLayouts` property should be an object that includes properties named
 
 ### `onChangeSelection`: `function`
 
-Callback that signals the user selected one of more items, and takes them as parameter. So far, only the `list` view implements it.
+Callback that signals the user selected one of more items, and takes them as parameter.
+
+### `isItemClickable`: `function`
```
Comment (youknowriad) [inline packages/dataviews/README.md:390]: Is this one really necessary? I mean what happens if we just do nothing in `onClickItem` instead?

Comment (gigitux) [reply packages/dataviews/README.md:390]: > Is this one really necessary? I mean what happens if we just do nothing in onClickItem instead?

This can be useful for differentiating the layout style of clickable primary fields from non-clickable ones. In the page view, the “primary” field is gray when the item is not clickable:


https://github.com/user-attachments/assets/2fdc2b6f-5a94-4710-ac37-4738dbc970ae


(also, I just noticed that `deleted` pages aren't visible in the `All pages` view)

File: packages/edit-site/src/components/post-list/style.scss [hunk 7]
```diff
@@ -64,6 +70,25 @@
 	overflow: hidden;
 }
 
+.dataviews-view-grid__primary-field.dataviews-view-grid__primary-field--clickable,
```
Comment (oandregal) [inline packages/edit-site/src/components/post-list/style.scss]: Can this style rule be moved to dataviews instead? This would be part of how a "clickable" field works as per the layout definition of a "clickable field". Consumers shouldn't have to deal with this.

File: packages/dataviews/src/dataviews-layouts/grid/index.tsx [hunk 8]
```diff
@@ -96,7 +121,27 @@ function GridItem< Item >( {
 				justify="space-between"
 				className="dataviews-view-grid__title-actions"
 			>
-				<HStack className="dataviews-view-grid__primary-field">
+				<HStack
```
Comment (oandregal) [inline packages/dataviews/src/dataviews-layouts/grid/index.tsx]: What do you think of following the same approach as the table layout? This is: adding an extra div for the content, so the click target is the title text instead of the "cell area":

https://github.com/user-attachments/assets/48d8b8f6-72b8-42d9-a9ed-8370360a4686

File: packages/dataviews/src/types.ts [hunk 9]
```diff
@@ -498,6 +498,8 @@ export interface ViewBaseProps< Item > {
 	onChangeSelection: SetSelection;
 	selection: string[];
 	setOpenedFilter: ( fieldId: string ) => void;
+	onItemClick: ( item: Item ) => void;
```
Comment (oandregal) [inline packages/dataviews/src/types.ts]: I'd prefer if we go with `onClickItem`, like the docs say — this would follow the existing naming schema (`onChangeSelection`, on + verb + noun).

**Comment by oandregal** [comment]
I am still [unconvinced](https://github.com/WordPress/gutenberg/pull/66365#issuecomment-2434694775) this approach would scale to other types of layouts with different interaction needs. It'd work as long as the only thing they need interaction wise is one click event, but we're back to square one as soon as the layout wants to support any other interaction.

It's also true that the layouts cannot be extended and have not yet a formalized API. I'm fine with shipping this and revisiting once that work happens.

File: packages/edit-site/src/components/post-fields/index.js (hunk 5 above)
Comment (louwie17) [reply packages/edit-site/src/components/post-fields/index.js:184]: Yeah, that's not ideal!
I am still leaning towards the fact that this should continue to be handled on a field level and maybe we only expose the `onItemClick` function that can be used as a callback with the `field` being clicked.

In order for us to do this, we did have to pass down a `onClick` function to the field render method.
Alluding to @oandregal [last thought](https://github.com/WordPress/gutenberg/pull/66365#issuecomment-2455329276), maybe we want to pass down a `onEvent` callback that takes the event type `click` or something else, and the `field`.
The render method within the field can decide how to handle this.

File: packages/edit-site/src/components/post-list/style.scss (hunk 7 above)
Comment (gigitux) [reply packages/edit-site/src/components/post-list/style.scss]: This CSS is specifically for the post title field. The only part that might make sense to move is the `cursor: pointer`, but I suggest keeping it here for now.

Comment (oandregal) [reply packages/edit-site/src/components/post-list/style.scss]: Shouldn't we formalise how "clickable" fields look like? Making the cursor a pointer, like any clickable element, seems something we should handle for the consumers.

If the other styles are only relevant for the title field, can we scope them using the `.edit-site-post-list__title`? We should aim for consumer code not to have any `.dataviews-*` CSS.

Comment (gigitux) [reply packages/edit-site/src/components/post-list/style.scss]: > Shouldn't we formalise how "clickable" fields look like? Making the cursor a pointer, like any clickable element, seems something we should handle for the consumers.

I'm happy to move the cursor rule.

> If the other styles are only relevant for the title field, can we scope them using the .edit-site-post-list__title? We should aim for consumer code not to have any .dataviews-* CSS.

Overall, I agree with this. The main issue is that this means that we should add `--clickable` class directly to the field.

File: packages/edit-site/src/components/post-fields/index.js [hunk 10]
```diff
@@ -139,7 +134,7 @@ function PostAuthorField( { item } ) {
 	);
 }
 
-function usePostFields( viewType ) {
+function usePostFields() {
```
Comment (youknowriad) [inline packages/edit-site/src/components/post-fields/index.js:137]: Nice change, it means we can probably move all these fields into the fields package from now on.

File: packages/dataviews/src/types.ts (hunk 9 above)
Comment (gigitux) [reply packages/dataviews/src/types.ts]: Fixed with 419f3fa7d2ee0da4c855087d7cc74ef88030cc4a

File: packages/dataviews/src/dataviews-layouts/grid/index.tsx (hunk 8 above)
Comment (gigitux) [reply packages/dataviews/src/dataviews-layouts/grid/index.tsx]: Fixed with 7a7964457a25010712721a454ee5d6b0c883ed63:

https://github.com/user-attachments/assets/670edb60-d237-47cf-a42c-065ee2aaf8a8

File: packages/dataviews/src/dataviews-layouts/grid/index.tsx [hunk 11]
```diff
@@ -81,7 +86,27 @@ function GridItem< Item >( {
...
 			} }
 		>
-			<div className="dataviews-view-grid__media">
+			<div
+				className={ clsx( 'dataviews-view-grid__media', {
+					'dataviews-view-grid__media--clickable':
+						isItemClickable( item ),
+				} ) }
+				tabIndex={ isItemClickable( item ) ? 0 : undefined }
+				role="button"
```
Comment (youknowriad) [inline packages/dataviews/src/dataviews-layouts/grid/index.tsx]: Should the "role" be "button" if the item is not clickable?

File: packages/dataviews/src/dataviews-layouts/grid/index.tsx [hunk 12]
```diff
@@ -81,7 +86,27 @@ function GridItem< Item >( {
...
+				onClick={ () => {
```
Comment (youknowriad) [inline packages/dataviews/src/dataviews-layouts/grid/index.tsx]: this prop should be undefined if the item is not clickable rather than checking within the callback.

File: packages/dataviews/src/dataviews-layouts/grid/index.tsx [hunk 13]
```diff
@@ -81,7 +86,27 @@ function GridItem< Item >( {
...
+					if ( isItemClickable( item ) ) {
+						onClickItem( item );
+					}
+				} }
+				onKeyDown={ ( event ) => {
```
Comment (youknowriad) [inline packages/dataviews/src/dataviews-layouts/grid/index.tsx]: same as above.

File: packages/dataviews/src/dataviews-layouts/table/index.tsx [hunk 14]
```diff
@@ -102,15 +110,40 @@ function TableColumnField< Item >( {
...
+					}
+				} }
+				onKeyDown={ ( event ) => {
+					if (
+						( event.key === 'Enter' || event.key === '' ) &&
+						isItemClickableField
+					) {
+						onClickItem( item );
+					}
+				} }
```
Comment (youknowriad) [inline packages/dataviews/src/dataviews-layouts/table/index.tsx]: It feels like we're repeating the same logic in all the places where we want to make something clickable. Maybe we can do have a reusable useClickableItemProps hook and reuse it in all these places:

```
const clickableProps = useClickableItemProps( item ); 
```

**Review by youknowriad — COMMENTED** [review]
Tested this and it works well for me.

File: packages/dataviews/src/dataviews-layouts/table/index.tsx (hunk 14 above)
Comment (gigitux) [reply packages/dataviews/src/dataviews-layouts/table/index.tsx]: Great idea! Thanks! 🙇 507d60b0db8dc4c056e3699a567394025d352d98

File: packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+			[ className + '--clickable' ]: isClickable,
+		} ),
+		role: isItemClickable( item ) ? 'button' : undefined,
+		onClick: () => {
+			if ( isClickable ) {
+				onClickItem( item );
+			}
+		},
+		tabIndex: isClickable ? 0 : undefined,
+		onKeyDown: ( event: React.KeyboardEvent ) => {
```
Comment (youknowriad) [inline packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]: Should make this `undefined` if isClickable is false. Same for onClick prop

File: packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+	onClickItem: ( item: Item ) => void,
+	className: string
+) => {
+	const isClickable = isItemClickable( item );
+
+	return {
+		className: clsx( className, {
+			[ className + '--clickable' ]: isClickable,
+		} ),
+		role: isItemClickable( item ) ? 'button' : undefined,
```
Comment (youknowriad) [inline packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]:
```suggestion
		role: isClickable ? 'button' : undefined,
```

**Review by youknowriad — COMMENTED** [review]
I want to acknowledge this is one possible approach, I know you folks are not entirely convinced. I'm not entirely ready to make this layout specific personally. 

Would you be ok with moving forward here and judging this over time?

File: packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts (hunk 15 above)
Comment (gigitux) [reply packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]: There is an if inside both functions to ensure that the callback is invoked only when the item is `clickable`.

https://github.com/WordPress/gutenberg/blob/507d60b0db8dc4c056e3699a567394025d352d98/packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts#L25-L32

https://github.com/WordPress/gutenberg/blob/507d60b0db8dc4c056e3699a567394025d352d98/packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts#L19-L23

I'm fine with replacing both with something like this if you prefer:

```ts
onKeyDown: ! isClickable
? undefined
: ( event: React.KeyboardEvent ) => {
		if ( event.key === 'Enter' || event.key === '' ) {
			onClickItem( item );
		}
  },

onClick: ! isClickable
? undefined
: () => {
		onClickItem( item );
},
```

**Comment by gigitux** [comment]
> I want to acknowledge this is one possible approach, I know you folks are not entirely convinced. I'm not entirely ready to make this layout specific personally.
> 
> Would you be ok with moving forward here and judging this over time?

The proposed approach works for me. Merging this PR will fix this regression: https://github.com/WordPress/gutenberg/issues/66588. We can continue to iterate as needed 💪

File: packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts (hunk 15 above)
Comment (youknowriad) [reply packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]: yeah I just thought avoiding the handlers on things that are not "button" would be best.

Comment (gigitux) [reply packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]: Fixed with c9e418b5fddc40020c9c677c3b29ae6dda162747!

File: packages/edit-site/src/components/post-list/style.scss [hunk 17]
```diff
@@ -64,6 +70,26 @@
 	overflow: hidden;
 }
 
+.dataviews-view-grid__primary-field.dataviews-view-grid__primary-field--clickable
```
Comment (oandregal) [inline packages/edit-site/src/components/post-list/style.scss:73]: We want these styles to be applied to the title _only_ if the layout supports the item being "clickable". Either we assume these classes become public API (so they should be stable) or we need to figure out better ways to do this.

**Review by oandregal — APPROVED** [review]
Thanks for your work on this, Luigi!

I find a bit problematic that consumers need to target [dataviews internal classes](https://github.com/WordPress/gutenberg/pull/66365/files#r1832171434) (so they become public API). This is an area I'd like us to iterate on.

Other than that, I think we should try this approach to gauge whether it's good enough.

File: packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts [hunk 18]
```diff
@@ -0,0 +1,29 @@
+/**
+ * External dependencies
+ */
+import clsx from 'clsx';
+
+export const useClickableItemProps = < Item >(
```
Comment (mcsf) [inline packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]: As it's currently implemented, this isn't a hook, so let's avoid the `use*` naming. Could be `get*` instead, or `compute*` to be even more unambiguous.

File: packages/dataviews/src/components/dataviews/index.tsx [hunk 19]
```diff
@@ -110,6 +114,8 @@ export default function DataViews< Item >( {
 				openedFilter,
 				setOpenedFilter,
 				getItemId,
+				isItemClickable: isItemClickable ?? ( () => false ),
+				onClickItem: onClickItem ?? ( () => {} ),
```
Comment (mcsf) [inline packages/dataviews/src/components/dataviews/index.tsx]: Should these be cached?

File: packages/dataviews/src/dataviews-layouts/grid/style.scss [hunk 20]
```diff
@@ -17,8 +17,13 @@
 
 		.dataviews-view-grid__primary-field {
 			min-height: $grid-unit-40; // Preserve layout when there is no ellipsis button
+
+			&--clickable {
+				width: fit-content;
```
Comment (mcsf) [inline packages/dataviews/src/dataviews-layouts/grid/style.scss:22]: Why is this needed, and why only for `--clickable`?

File: packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+		tabIndex: isClickable ? 0 : undefined,
+		onClick: ! isClickable ? undefined : () => onClickItem( item ),
+		onKeyDown: ! isClickable
+			? undefined
+			: ( event: React.KeyboardEvent ) => {
+					if ( event.key === 'Enter' || event.key === '' ) {
+						onClickItem( item );
+					}
+			  },
+	};
```
Comment (mcsf) [inline packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]: This seems needlessly complicated. If the item isn't clickable, all attributes but class will be empty. So:

```suggestion
	if ( ! isItemClickable( item ) ) {
		return { className };
	}

	return {
		className: `${ className } ${ className }--clickable`,
		role: 'button',
		tabIndex: 0,
		onClick: () => onClickItem( item ),
		onKeyDown: ( event: React.KeyboardEvent ) => {
			if ( event.key === 'Enter' || event.key === '' ) {
				onClickItem( item );
			}
		},
	};
```

Also, see my other comment about `&--clickable`. If it turns out we don't need that class, then this could be further simplified by not handling `className` altogether. The caller would handle its own class more explicitly:

```js
const clickableProps = useClickableItemProps(
	item,
	isItemClickableField,
	onClickItem
);

return (
	<div
		className="dataviews-view-table__cell-content"
		{ ...clickableProps }
```

File: packages/dataviews/src/dataviews-layouts/grid/style.scss [hunk 22]
```diff
@@ -17,8 +17,13 @@
 
 		.dataviews-view-grid__primary-field {
 			min-height: $grid-unit-40; // Preserve layout when there is no ellipsis button
+
+			&--clickable {
```
Comment (mcsf) [inline packages/dataviews/src/dataviews-layouts/grid/style.scss:21]: Is `&--clickable` the best way to achieve this? We don't seem to do this in Gutenberg (we have only one occurrence of `&--is-hex`). What does this do that couldn't be achieved by targeting `&[role="button"]`?

File: packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts (hunk 21 above)
Comment (oandregal) [reply packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]: Addressed at https://github.com/WordPress/gutenberg/pull/66365/commits/f24172ba41318288889fb46fe1689f90fbdc74ff

File: packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts (hunk 18 above)
Comment (oandregal) [reply packages/dataviews/src/dataviews-layouts/hooks/use-clickable-item-props.ts]: Addressed at https://github.com/WordPress/gutenberg/pull/66365/commits/fdc2892f34acae7e6230ff6d7de1f0fa68240576

File: packages/dataviews/src/components/dataviews/index.tsx (hunk 19 above)
Comment (oandregal) [reply packages/dataviews/src/components/dataviews/index.tsx]: Addressed at https://github.com/WordPress/gutenberg/pull/66365/commits/84cc4c27e220e105d2ddb776d0808039f1df966f

File: packages/dataviews/src/dataviews-layouts/grid/style.scss (hunk 22 above)
Comment (oandregal) [reply packages/dataviews/src/dataviews-layouts/grid/style.scss:21]: I agree the way this handles the classes needs to be iterated, see related conversation https://github.com/WordPress/gutenberg/pull/66365#discussion_r1832171434

In discussing with Riad IRL the idea was to land this PR and iterate.

**Comment by oandregal** [comment]
I want to build on top of this PR, so I've pushed changes to address @mcsf feedback. This PR is not perfect and we need to iterate, but I'm going to merge to unblock related work.

**Comment by mcsf** [comment]
> I want to build on top of this PR, so I've pushed changes to address @mcsf feedback. This PR is not perfect and we need to iterate, but I'm going to merge to unblock related work.

Thanks for addressing most of it, @oandregal! I'm quite fine with merging it as is. Let's address the rest afterwards.
